Do not check object references for equality after hot reload

buyaa-n opened the pull request:
RuntimeType member's cache is purged after hot reload, therefore same member references are not equal after hotreload which causing issue like mentioned in [Hot Reload breaks MethodInfo equality](https://github.com/dotnet/runtime/issues/69427) 

Updated/added Object.Equals overloads to compare metadata token and module in case hot reload taken place instead of comparing object references.

Fixes https://github.com/dotnet/runtime/issues/69427

buyaa-n committed:
Equality check of type members after hot reload

buyaa-n commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:
```diff
@@ -168,44 +169,51 @@ public override int GetHashCode()
         {
-            if (!IsGenericMethod)
-                return obj == (object)this;
+            if (IsGenericMethod)
```
There is nothing changed within IsGenericMethod true case, just reverted the conditional

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeFieldInfo.cs:
```diff
@@ -46,6 +47,12 @@ internal RuntimeType GetDeclaringTypeInternal()
+            obj == (object)this ||
+            (RuntimeTypeMetadataUpdateHandler.HotReloadDeltaApplied && CacheEquals(obj));
+
+        public override int GetHashCode() => base.GetHashCode();
```
I do not think you can just forward to the `base.GetHashCode()`. `GetHashCode` and `Equals` have to be in sync. It means `GetHashCode` needs to have `HotReloadDeltaApplied` condition too.

jkotas replied:
Also, add tests that verify that the hashcodes are Equal even when instances are different.

buyaa-n replied:
I felt the same, but was not sure how hot reload affects GetHashCode, probably add tests to see the diff

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/RuntimeTypeMetadataUpdateHandler.cs:
```diff
@@ -11,11 +11,14 @@ namespace System.Reflection.Metadata
     /// <summary>Metadata update handler used to clear a Type's reflection cache in response to a metadata update notification.</summary>
     internal static class RuntimeTypeMetadataUpdateHandler
     {
+        public static bool HotReloadDeltaApplied { get; internal set; }
```
We should make this unconditionally return false when hot reload support is stripped.

jkotas replied:
Look for System.Reflection.Metadata.MetadataUpdater.IsSupported in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml for how to do it.

buyaa-n replied:
> We should make this unconditionally return false when hot reload support is stripped.

Isn't it should also apply to the ClearCache(...) too? Do not clear cache if it is not supported.

jkotas replied:
The ClearCache method should be taken care of by this already: https://github.com/dotnet/runtime/blob/b4991a5a43d379f271e1f26286eb24709e62da9c/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml#L33-L37

When hot reload is not supported, the MetadataUpdateHandlerAttribute will be deleted. It means that nothing is going to be keeping the RuntimeTypeMetadataUpdateHandler alive and the whole thing should be trimmed from the final binary.

buyaa-n replied:
The HotReloadDeltaApplied is false by default and it is set true only if ClearCache() called, by default it is false (I will update the setter to private), then do we still need to check if hot reload supported each time?

jkotas replied:
Trimming is only able to do something with constants. Static variable is not a constant, even when it is never assigned to. Trimming is not able to figure out that a static variable that is never assigned to is a constant in practice.

buyaa-n replied:
It helps thanks, I was about to check `MetadataUpdater.IsSupported` each time `HotReloadDeltaApplied` is referenced 😛

buyaa-n replied:
Update: in offline discussion @jkotas suggested its better to `just check MetadataUpdater.IsSupported. It has larger scope - it will avoid the perf overhead even when trimming is not used (ie majority of our customers)`

jkotas commented on src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs:
```diff
@@ -60,5 +118,13 @@ private static Action<Type[]> GetClearCacheMethod()
+        private static bool HotReloadDeltaApplied()
+        {
+            Type updateHandler = typeof(Type).Assembly.GetType("System.Reflection.Metadata.RuntimeTypeMetadataUpdateHandler", throwOnError: true, ignoreCase: false);
+            PropertyInfo hotReloadApplied = updateHandler.GetProperty("HotReloadDeltaApplied", BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static);
```
Do we really need to be testing this internal implementation detail?

buyaa-n replied:
not necessarily

AaronRobinsonMSFT commented on src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs:
```diff
@@ -40,17 +68,47 @@ public void GetMethod_MultipleCalls_ClearCache_DifferentObjects(bool justSpecifi
+            Assert.Equal(nameof(Field1), fi2.Name);
 
             Assert.NotSame(mi1, mi2);
+            Assert.NotSame(pi1, pi2);
```
This

AaronRobinsonMSFT commented on src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs:
```diff
@@ -40,17 +68,47 @@ public void GetMethod_MultipleCalls_ClearCache_DifferentObjects(bool justSpecifi
+            Assert.Equal(nameof(Property), pi2.Name);
+            Assert.Equal(nameof(Field1), fi2.Name);
 
             Assert.NotSame(mi1, mi2);
```
This block of `NotSame` and `Equal` is different enough that I think it warrants a comment above the block inidcating what we are checking.

buyaa-n replied:
Right

buyaa-n committed:
Apply comments

stephentoub commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/RuntimeTypeMetadataUpdateHandler.cs:
```diff
@@ -11,11 +11,15 @@ namespace System.Reflection.Metadata
     /// <summary>Metadata update handler used to clear a Type's reflection cache in response to a metadata update notification.</summary>
     internal static class RuntimeTypeMetadataUpdateHandler
     {
+        public static bool HotReloadDeltaApplied { get; private set; }
```
Could we make the name of this more generic and in line with the method that actually sets it, e.g. instead of `HotReloadDeltaApplied`, call it `CachesCleared` or something like that?

buyaa-n replied:
Sure

stephentoub commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeEventInfo.cs:
```diff
@@ -68,6 +69,12 @@ public override string ToString()
+            obj == (object)this ||
+            (RuntimeTypeMetadataUpdateHandler.HotReloadDeltaApplied && CacheEquals(obj));
+
+        public override int GetHashCode() => HashCode.Combine(m_token.GetHashCode(), RuntimeTypeHandle.GetModule(m_declaringType).GetHashCode());
```
Is this going to be more expensive than it is today in the case where a delta hasn't been applied?

buyaa-n replied:
I will run some perf tests

buyaa-n replied:
The result looks bad, maybe should and call the old method if the Cache is not cleared 

|                   Method |        Job |         Toolchain |      Mean |    Error |   StdDev | Ratio | RatioSD |  Gen 0 | Allocated |
|------------------------- |----------- |------------------ |----------:|---------:|---------:|------:|--------:|-------:|----------:|
|       GetAMethodHashCode | Job-ARUJVB | \main\corerun.exe |  32.57 ns | 0.207 ns | 0.194 ns |  1.00 |    0.00 |      - |         - |
|       GetAMethodHashCode | Job-FDOPUD |   \pr\corerun.exe |  33.09 ns | 0.264 ns | 0.247 ns |  1.02 |    0.01 |      - |         - |
|                          |            |                   |           |          |          |       |         |        |           |
|    GetAllMethodsHashCode | Job-ARUJVB | \main\corerun.exe | 665.92 ns | 4.653 ns | 3.885 ns |  1.00 |    0.00 | 0.1259 |     528 B |
|    GetAllMethodsHashCode | Job-FDOPUD |   \pr\corerun.exe | 681.18 ns | 7.214 ns | 6.748 ns |  1.02 |    0.01 | 0.1259 |     528 B |
|                          |            |                   |           |          |          |       |         |        |           |
|     GetAPropertyHashCode | Job-ARUJVB | \main\corerun.exe |  32.31 ns | 0.226 ns | 0.201 ns |  1.00 |    0.00 |      - |         - |
|     GetAPropertyHashCode | Job-FDOPUD |   \pr\corerun.exe |  48.84 ns | 0.665 ns | 0.589 ns |  1.51 |    0.02 |      - |         - |
|                          |            |                   |           |          |          |       |         |        |           |
| GetAllPropertiesHashCode | Job-ARUJVB | \main\corerun.exe | 112.52 ns | 0.280 ns | 0.249 ns |  1.00 |    0.00 | 0.0191 |      80 B |
| GetAllPropertiesHashCode | Job-FDOPUD |   \pr\corerun.exe | 194.13 ns | 1.908 ns | 1.594 ns |  1.73 |    0.02 | 0.0191 |      80 B |
|                          |            |                   |           |          |          |       |         |        |           |
|        GetAFieldHashCode | Job-ARUJVB | \main\corerun.exe |  20.09 ns | 0.158 ns | 0.140 ns |  1.00 |    0.00 |      - |         - |
|        GetAFieldHashCode | Job-FDOPUD |   \pr\corerun.exe |  47.11 ns | 0.332 ns | 0.311 ns |  2.35 |    0.02 |      - |         - |
|                          |            |                   |           |          |          |       |         |        |           |
|     GetAllFieldsHashCode | Job-ARUJVB | \main\corerun.exe | 135.10 ns | 2.511 ns | 3.909 ns |  1.00 |    0.00 | 0.0439 |     184 B |
|     GetAllFieldsHashCode | Job-FDOPUD |   \pr\corerun.exe | 263.37 ns | 1.620 ns | 1.352 ns |  1.92 |    0.06 | 0.0439 |     184 B |

buyaa-n replied:
After adding `MetadataUpdater` check when it is not supported the results looks comparable:

|                   Method |        Job |         Toolchain |      Mean |     Error |     StdDev | Ratio | RatioSD |  Gen 0 | Allocated |
|------------------------- |----------- |------------------ |----------:|----------:|-----------:|------:|--------:|-------:|----------:|
|     GetAPropertyHashCode | Job-CBGIOV | \main\corerun.exe |  34.69 ns |  0.526 ns |   0.492 ns |  1.00 |    0.00 |      - |         - |
|     GetAPropertyHashCode | Job-IRZIMW |   \pr\corerun.exe |  33.30 ns |  0.583 ns |   0.455 ns |  0.96 |    0.02 |      - |         - |
|                          |            |                   |           |           |            |       |         |        |           |
|        GetAFieldHashCode | Job-CBGIOV | \main\corerun.exe |  21.22 ns |  0.225 ns |   0.200 ns |  1.00 |    0.00 |      - |         - |
|        GetAFieldHashCode | Job-IRZIMW |   \pr\corerun.exe |  21.85 ns |  0.195 ns |   0.173 ns |  1.03 |    0.01 |      - |         - |

When `MetadataUpdater` is supported after updating the hashcode logic to `just combine the hashcode of the type and the token` as @jkotas suggested perf also improved 

|               Method |        Job |         Toolchain |     Mean |    Error |   StdDev | Ratio | RatioSD | Allocated |
|--------------------- |----------- |------------------ |---------:|---------:|---------:|------:|--------:|----------:|
| GetAPropertyHashCode | Job-QHEWUL | \main\corerun.exe | 32.41 ns | 0.273 ns | 0.255 ns |  1.00 |    0.00 |         - |
| GetAPropertyHashCode | Job-FJTZYE |   \pr\corerun.exe | 36.17 ns | 0.305 ns | 0.255 ns |  1.12 |    0.01 |         - |
|                      |            |                   |          |          |          |       |         |           |
|    GetAFieldHashCode | Job-QHEWUL | \main\corerun.exe | 20.04 ns | 0.283 ns | 0.236 ns |  1.00 |    0.00 |         - |
|    GetAFieldHashCode | Job-FJTZYE |   \pr\corerun.exe | 27.57 ns | 0.131 ns | 0.116 ns |  1.38 |    0.02 |         - |
|                      |            |                   |          |          |          |       |         |           |
|   GetAnEventHashCode | Job-QHEWUL | \main\corerun.exe | 20.31 ns | 0.136 ns | 0.120 ns |  1.00 |    0.00 |         - |
|   GetAnEventHashCode | Job-FJTZYE |   \pr\corerun.exe | 23.76 ns | 0.395 ns | 0.369 ns |  1.17 |    0.02 |         - |

buyaa-n committed:
Apply feedback

buyaa-n commented on src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs:
```diff
@@ -33,24 +76,125 @@ public void InvokeClearCache_NoExceptions()
         }
 
         [ActiveIssue("https://github.com/dotnet/runtime/issues/50978", TestRuntimes.Mono)]
-        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsMetadataUpdateSupported))]
```
Changed the `Theory` to `Fact` because of getting exception with `RemoteExecutor` and could not find how to fix
```log
Assert.All() Failure: 1 out of 1 items in the collection did not pass.
[0]: Item: Boolean justSpecificType
     Xunit.Sdk.TrueException: Field marshaling is not supported by Invoke: justSpecificType
     Expected: True
     Actual:   False
        at Xunit.Assert.True(Nullable`1 condition, String userMessage) in /_/src/xunit.assert/Asserts/BooleanAsserts.cs:line 132
        at Xunit.Assert.True(Boolean condition, String userMessage) in /_/src/xunit.assert/Asserts/BooleanAsserts.cs:line 116
        at Microsoft.DotNet.RemoteExecutor.RemoteExecutor.<>c.<GetMethodInfo>b__37_0(FieldInfo fi) in /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteExecutor.cs:line 504
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/xunit.assert/Asserts/CollectionAsserts.cs:line 34
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/xunit.assert/Asserts/CollectionAsserts.cs:line 61
```

buyaa-n committed:
Renamed the propertyName in test

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeEventInfo.cs:
```diff
@@ -68,6 +69,12 @@ public override string ToString()
+            obj == (object)this ||
+            (RuntimeTypeMetadataUpdateHandler.MetadataUpdaterSupportedAndCacheCleared && CacheEquals(obj));
+
+        public override int GetHashCode() => HashCode.Combine(m_token.GetHashCode(), RuntimeTypeHandle.GetModule(m_declaringType).GetHashCode());
```
You can just combine the hashcode of the type and the token. You do not need to go to the module.

jkotas replied:
Similarly, Equals can just compare the type.

buyaa-n replied:
Can apply for all other members?

jkotas replied:
I think so. Also, the cache equals method can use that too.

buyaa-n committed:
Compare type instead of module, check MetadataUpdater.IsSupported in HashCode

buyaa-n committed:
Update and use CacheEquals

stephentoub commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/Metadata/RuntimeTypeMetadataUpdateHandler.cs:
```diff
@@ -11,11 +11,16 @@ namespace System.Reflection.Metadata
     internal static class RuntimeTypeMetadataUpdateHandler
     {
+        private static bool s_cacheCleared;
+        public static bool UpdateSupportedAndCacheCleared => MetadataUpdater.IsSupported && s_cacheCleared;
```
Why is the IsSupported check needed here?  Is that to help with trimming?

If it's for trimming, ok.  Otherwise, s_cacheCleared would never be set to true if !IsSupported.

jkotas replied:
It is to help with both trimming and with tiering compilation. 

`MetadataUpdater.IsSupported` is readonly and so the tiered JIT is able to dead-code eliminate the whole check and make this change zero perf impact when hot-reload is not supported.

stephentoub replied:
Makes sense.

stephentoub replied:
I wonder then if we're making this too complicated.  Should we just always use the slightly more expensive Equals/GetHashCode implementations if MetadataUpdater.IsSupported is true, and skip this cache clearing tracking?  I don't have a strong opinion, but there are plenty of other expenses when debugging, which is when MetadataUpdater.IsSupported would be true.

jkotas replied:
Yes, I think it would make sense.

buyaa-n replied:
If the cache never cleared the equals could be a bit faster, but there is not way to Reset it back, so after one hot reload it would always use slower method, so I am OK with removing this property

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:
```diff
@@ -163,49 +164,56 @@ public override int GetHashCode()
             if (IsGenericMethod)
                 return ValueType.GetHashCodeOfPtr(m_handle);
             else
-                return base.GetHashCode();
+                return m_handle.GetHashCode();
```
This should return `m_handle.GetHashCodeOfPtr(m_handle)` in all cases. Also all other places that are hashing the handle should use `GetHashCodeOfPtr` instead of `.GetHashCode()`

jkotas replied:
`GetHashCodeOfPtr` does less work and it should be faster than `base.GetHashCode()`. If it is not faster, we should tweak the implementation to make it master.

buyaa-n replied:
sure, I noticed `m_handle.GetHashCode()` was also bit faster than `base.GetHashCode()`, but apparently might have been producing it wrong

steveharter commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.CoreCLR.cs:
```diff
@@ -118,6 +119,15 @@ public override string ToString()
         }
+
+        public override bool Equals(object? obj) =>
+            obj == (object)this ||
```
nit: for clarity instead of `==` which can be overloaded, use `object.ReferenceEquals(this, obj)`

steveharter commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeEventInfo.cs:
```diff
@@ -53,8 +54,7 @@ internal override bool CacheEquals(object? o)
                 m.m_token == m_token &&
-                RuntimeTypeHandle.GetModule(m_declaringType).Equals(
-                    RuntimeTypeHandle.GetModule(m.m_declaringType));
+                m_declaringType.Equals(m.m_declaringType);
```
Was `Type.Equals` choosen over Module equality since it's faster? Or is it a correctness thing?

buyaa-n replied:
I believe because it is faster, also assuming correctness also satisfied as .Net Core support only one module types are unique within a module

steveharter commented on src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs:
```diff
@@ -2,23 +2,66 @@
+        private static bool IsMetadataUpdateAndRemoteExecutorSupported => PlatformDetection.IsMetadataUpdateSupported && RemoteExecutor.IsSupported;
+
+        private static readonly Type s_type = typeof(ReflectionCacheTests);
+        //private BindingFlags flags = BindingFlags.NonPublic | BindingFlags.Instance;
```
nit: commented out code

>//private BindingFlags flags = BindingFlags.NonPublic | BindingFlags.Instance;

buyaa-n committed:
Apply feedbacks

buyaa-n committed:
Use ReferenceEquals

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeEventInfo.cs:
```diff
@@ -68,6 +68,13 @@ public override string ToString()
         }
+
+        public override bool Equals(object? obj) =>
+            obj == (object)this ||
```
Same as change to `ReferenceEquals` as https://github.com/dotnet/runtime/pull/73009/files#r935802561

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeFieldInfo.cs:
```diff
@@ -46,6 +47,13 @@ internal RuntimeType GetDeclaringTypeInternal()
         public override bool IsCollectible => m_declaringType.IsCollectible;
+
+        public override bool Equals(object? obj) =>
+            obj == (object)this ||
```
Same as change to `ReferenceEquals` as https://github.com/dotnet/runtime/pull/73009/files#r935802561

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:
```diff
@@ -157,55 +158,55 @@ public override string ToString()
-            for (int i = 0; i < lhs.Length; i++)
-            {
-                if (lhs[i] != rhs[i])
+                if (DeclaringType != mi.DeclaringType)
```
Wondering if the `DeclaringType` and `ReflectedType` comparisons should be moved up before the generic argument checks. Likely to be cheaper.

buyaa-n replied:
There is nothing changed within this `if (IsGenericMethod)` conditional, just indentation change, I don't really want to touch unrelated code

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:
```diff
@@ -157,55 +158,55 @@ public override string ToString()
+                return true;
+            }
+
+            return obj == (object)this ||
```
Probably as change to `ReferenceEquals` as https://github.com/dotnet/runtime/pull/73009/files#r935802561

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:
```diff
@@ -157,55 +158,55 @@ public override string ToString()
+
+            return obj == (object)this ||
+                (MetadataUpdater.IsSupported &&
+                    obj is RuntimeMethodInfo m &&
```
We do the coercing to `RuntimeMethodInfo` twice here in 206 and at 171. Is there a clearer path where we only do that once?

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeParameterInfo.cs:
```diff
@@ -496,6 +497,15 @@ public override Type[] GetOptionalCustomModifiers()
         }
 
+        public override bool Equals(object? obj) =>
+            obj == (object)this ||
```
Same as change to `ReferenceEquals` as https://github.com/dotnet/runtime/pull/73009/files#r935802561

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimePropertyInfo.cs:
```diff
@@ -179,6 +179,13 @@ public override IList<CustomAttributeData> GetCustomAttributesData()
         public override bool IsCollectible => m_declaringType.IsCollectible;
+
+        public override bool Equals(object? obj) =>
+            obj == (object)this ||
```
Same as change to `ReferenceEquals` as https://github.com/dotnet/runtime/pull/73009/files#r935802561

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/MdFieldInfo.cs:
```diff
@@ -33,8 +33,7 @@ internal override bool CacheEquals(object? o)
                 m.m_tkField == m_tkField &&
-                m_declaringType.TypeHandle.GetModuleHandle().Equals(
-                    m.m_declaringType.TypeHandle.GetModuleHandle());
+                m_declaringType.Equals(m.m_declaringType);
```
This should be also ReferenceEquals. Equals is virtual method - I am not sure whether the JIT is going to be able to de-virtualize it.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.CoreCLR.cs:
```diff
@@ -118,6 +119,15 @@ public override string ToString()
         }
+
+        public override bool Equals(object? obj) =>
+            object.ReferenceEquals(this, obj) ||
```
```suggestion
            ReferenceEquals(this, obj) ||
```
`object.` is unnecessary

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.CoreCLR.cs:
```diff
@@ -118,6 +119,15 @@ public override string ToString()
+                MetadataToken == ci.MetadataToken &&
+                m_declaringType.Equals(ci.m_declaringType));
+
+        public override int GetHashCode() => ValueType.GetHashCodeOfPtr(m_handle);
```
You should merge from main to resolve merge conflicts and fix build breaks. `GetHashCodeOfPtr` was moved recently.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:
```diff
@@ -157,55 +158,55 @@ public override string ToString()
         {
-            if (!IsGenericMethod)
-                return obj == (object)this;
+            if (IsGenericMethod)
```
I think that this is way overengineered for what it does. You should be able to change it to just:
```
public override bool Equals(object? obj)
{
    // We cannot do simple object identity comparisons for generic methods.
    // Equals will be called in CerHashTable when RuntimeType+RuntimeTypeCache.GetGenericMethodInfo()
    // retrieve items from and insert items into s_methodInstantiations which is a 
    // CerHashtable.
    return obj is RuntimeMethodInfo m && m_handle == m.m_handle;
}
```

buyaa-n committed:
Merge conflicts

buyaa-n committed:
apply more feedback

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeParameterInfo.cs:
```diff
@@ -496,6 +497,15 @@ public override Type[] GetOptionalCustomModifiers()
+                (MetadataUpdater.IsSupported &&
+                    obj is RuntimeParameterInfo pi &&
+                    m_tkParamDef == pi.m_tkParamDef &&
+                    ReferenceEquals(DefiningMethod.DeclaringType, pi.DefiningMethod.DeclaringType));
```
Whis does not work for global methods. DeclaringType is going to be null for them, and so we may incorrectly return that two ParameterInfos are equal even though they are from two unrelated modules.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeParameterInfo.cs:
```diff
@@ -496,6 +497,15 @@ public override Type[] GetOptionalCustomModifiers()
                 m_signature.GetCustomModifiers(PositionImpl + 1, false);
         }
 
+        public override bool Equals(object? obj) =>
```
Do we really need to fix ParameterInfo? 

It is unlikely that ParameterInfo is used as a key anywhere. We are not maintaining strong referential equality for ParameterInfo today. For example, the intialization at https://github.com/dotnet/runtime/blob/0ba3d3124e1ba3bc5be676ca33eaca0868858dd5/src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs#L80 is not thread safe - two threads may get two different instances of ParameterInfo.

buyaa-n replied:
not sure, just trying to cover all `*Infos`, will revert then

buyaa-n committed:
Exclude RuntimeParameterInfo

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/MdFieldInfo.cs:
```diff
@@ -33,8 +33,7 @@ internal override bool CacheEquals(object? o)
                 m.m_tkField == m_tkField &&
-                m_declaringType.TypeHandle.GetModuleHandle().Equals(
-                    m.m_declaringType.TypeHandle.GetModuleHandle());
+                ReferenceEquals(m_declaringType, m.m_declaringType);
```
Does the `MdFieldInfo` needs overridden GetHashCode and Equals? It is part of the same caches as RuntimeFieldInfo, so I think it should.

buyaa-n replied:
`RuntimeFieldInfo` is the parent of `MdFieldInfo` and `RtFieldInfo` is that necessary?

jkotas replied:
I have missed this detail. Ignore this comment.

buyaa-n replied:
`RtFieldInfo.CacheEquals` compares handles, so probably same as MethodInfo not needed
       internal override bool CacheEquals(object? o)
        {
            return o is RtFieldInfo m && m.m_fieldHandle == m_fieldHandle;
        }

jkotas commented on src/libraries/System.Runtime/tests/System/Reflection/ReflectionCacheTests.cs:
```diff
@@ -33,24 +71,116 @@ public void InvokeClearCache_NoExceptions()
+            Assert.Equal(ci1Hash, ci2Hash);
+        }
 
+        private static void AssertMemberReferencesNotSameButEqual(MethodInfo mi1, PropertyInfo pi1, FieldInfo fi1, EventInfo ei1, ConstructorInfo ci1,
```
Nit: These methods with a lot of arguments (in particular this one with 10 arguments) do look nice. Would it be better to have a method like this:

```
void AssertNotSameSameButEqual(object o1, object o2)
{
     Assert.NotSame(o1, o2);
     Assert.Equal(o1, o2);
     Assert.Equal(o1.GetHashCode(), o2.GetHashCode());
}
```
and call it on the individual pairs?

buyaa-n replied:
Sure

buyaa-n committed:
Update assert methods in test

jkotas reviewed (approved):
LGTM. Thank you!

buyaa-n commented:
CI test failures on `Microsoft.CSharp.RuntimeBinder.Tests`, `System.Dynamic.Runtime.Tests`, `System.Composition.Convention.Tests` could be related, I can repro locally and seems not reproduceable in other branch, looking

buyaa-n committed:
Update RMethodInfo RConstructorInfo Equals override to fix CI test failures

buyaa-n commented:
Comparing `RuntimeMethodInfo` handles for equality causing the CI test failures in above projects, reverted that

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:
```diff
@@ -157,56 +158,14 @@ public override string ToString()
-
-            return true;
-        }
+        public override bool Equals(object? obj) =>
```
This won't work with the `CerHashTable`

jkotas reviewed (changes requested):
> Comparing RuntimeMethodInfo handles for equality causing the CI test failures in above projects, reverted that

Do you understand why? Have you been able to create a small standalone repro using just the reflection APIs?

I would like to understand why it does not work. If this does not work, there may be more problems with this change.

buyaa-n commented:
> Do you understand why? Have you been able to create a small standalone repro using just the reflection APIs?

No, I could not understood why, a typical failure log looks like this:

```log
Dynamic.Operator.Tests.PlusEqualTypeTests.Byte [FAIL]
System.ArgumentException : ParameterExpression of type 'System.String' cannot be used for delegate parameter of type 'System.Object'
        Stack Trace:
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(954,0): at System.Linq.Expressions.Expression.ValidateLambdaArgs(Type delegateType, Expres
  sion& body, ReadOnlyCollection`1 parameters, String paramName)
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(717,0): at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, String na
  me, Boolean tailCall, IEnumerable`1 parameters)
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Runtime\CompilerServices\CallSiteBinder.cs(192,0): at System.Runtime.CompilerServices.CallSiteBinder.Stitch[T](Expression bindi
  ng, LambdaSignature`1 signature)
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Runtime\CompilerServices\CallSiteBinder.cs(132,0): at System.Runtime.CompilerServices.CallSiteBinder.BindCore[T](CallSite`1 sit
  e, Object[] args)
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Dynamic\UpdateDelegates.Generated.cs(157,0): at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0
  arg0, T1 arg1)
          D:\dotnet\pr2\src\libraries\System.Dynamic.Runtime\tests\Dynamic.Context\Conformance.dynamic.context.operator.compound.type.PlusEqual.cs(76,0): at Dynamic.Operator.Tests.PlusEqualTypeTests.B
  yte()
             at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
          D:\dotnet\pr2\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodInvoker.cs(69,0): at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```
just know reverting back the Equals logic fixes those failures

buyaa-n commented:
Another log
```log
System.Text.Tests.EncodingTests.GetEncoding_FromProvider_ByCodePage_WithDisallowedEncoding_Throws(encodingName: "utf-7", codePage: 65000) [FAIL]
        System.ArgumentException : ParameterExpression of type 'System.Text.EncodingProvider' cannot be used for delegate parameter of type 'System.Reflection.Binder'
        Stack Trace:
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(954,0): at System.Linq.Expressions.Expression.ValidateLambdaArgs(Type delegateType, Expres
  sion& body, ReadOnlyCollection`1 parameters, String paramName)
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(717,0): at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, String na
  me, Boolean tailCall, IEnumerable`1 parameters)
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(689,0): at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, Boolean t
  ailCall, IEnumerable`1 parameters)
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(651,0): at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, Parameter
  Expression[] parameters)
          D:\dotnet\pr2\src\libraries\System.Runtime\tests\System\Text\EncodingTests.cs(43,0): at System.Text.Tests.EncodingTests.GetEncoding_FromProvider_ByCodePage_WithDisallowedEncoding_Throws(Stri
  ng encodingName, Int32 codePage)
             at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
          D:\dotnet\pr2\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodInvoker.cs(64,0): at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
        System.ArgumentException : ParameterExpression of type 'System.Text.EncodingProvider' cannot be used for delegate parameter of type 'System.Reflection.Binder'
        Stack Trace:
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(954,0): at System.Linq.Expressions.Expression.ValidateLambdaArgs(Type delegateType, Expres
  sion& body, ReadOnlyCollection`1 parameters, String paramName)
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(717,0): at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, String na
  me, Boolean tailCall, IEnumerable`1 parameters)
      System.Text.Tests.EncodingTests.GetEncoding_FromProvider_ByEncodingName_WithDisallowedEncoding_Throws(encodingName: "utf-7", codePage: 65000) [FAIL]
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(689,0): at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, Boolean t
  ailCall, IEnumerable`1 parameters)
          D:\dotnet\pr2\src\libraries\System.Linq.Expressions\src\System\Linq\Expressions\LambdaExpression.cs(651,0): at System.Linq.Expressions.Expression.Lambda[TDelegate](Expression body, Parameter
  Expression[] parameters)
          D:\dotnet\pr2\src\libraries\System.Runtime\tests\System\Text\EncodingTests.cs(71,0): at System.Text.Tests.EncodingTests.GetEncoding_FromProvider_ByEncodingName_WithDisallowedEncoding_Throws(
  String encodingName, Int32 codePage)
             at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
          D:\dotnet\pr2\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodInvoker.cs(64,0): at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

jkotas committed:
Revert "Update RMethodInfo RConstructorInfo Equals override to fix CI test failures"

This reverts commit d4d961c81db83df4e9631c02782791baef4c6eea.

jkotas committed:
Fix tests

buyaa-n committed:
Update RuntimeConstructorInfo Equals, HashCode similarly

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeFieldInfo.cs:
```diff
@@ -46,6 +47,13 @@ internal RuntimeType GetDeclaringTypeInternal()
         public override Module Module => GetRuntimeModule();
         public override bool IsCollectible => m_declaringType.IsCollectible;
+
+        public override bool Equals(object? obj) =>
```
We end up not having any specialization for MethodInfo. MethodInfo is the most frequently used type among the ones modified in this PR.

I am not sure whether `MetadataUpdater.IsSupported` specialization for RuntimeFieldInfo, RuntimePropertyInfo and RuntimeEventInfo - that are all less important than RuntimeMethodInfo - is worth the trouble.

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.CoreCLR.cs:
```diff
@@ -118,6 +119,11 @@ public override string ToString()
         }
+
+        public override bool Equals(object? obj) =>
+            obj is RuntimeConstructorInfo ci && m_handle == ci.m_handle && ReferenceEquals(m_declaringType, ci.m_declaringType);
```
Probably should have a `ReferenceEquals(this, obj) ||` at the beginning.

buyaa-n replied:
Not sure if that would help with perf, need to test

jkotas replied:
I think that the perf of the two options is close enough. There is a value in simplicity too.

buyaa-n replied:
> I think that the perf of the two options is close enough

Yes there were no noticeable difference, but GetHashCode was quite slow, so added `MetadataUpdater.IsSupported` to conditionally use the old approach `base.GetHashCode()`

IDisposable commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:
```diff
@@ -157,56 +158,12 @@ public override string ToString()
-            Type[] lhs = GetGenericArguments();
-            Type[] rhs = mi.GetGenericArguments();
+        public override bool Equals(object? obj) =>
+            obj is RuntimeMethodInfo m && m_handle == m.m_handle && ReferenceEquals(m_declaringType, m.m_declaringType);
```
Probably should have a `ReferenceEquals(this, obj) ||` at the beginning.

jkotas commented:
Looking at the test failures - the reflected type needs to be also included in Equals method.

buyaa-n committed:
Add reflected type into Equals, check MetadataUpdater.IsSupported in GetHashCode

jkotas committed:
Perf tweaks for Equals and GetHashCode

- No need to use GetHashCodeOfPtr when we are wrapping the hashcode using HashCode.Combine that takes care of the randomization
- Use underlying type handle for type hashcodes. It is faster than the default object hashcode.
- Delete special casing for MetadataUpdater.IsSupported from RuntimeMethodInfo.GetHashCode. It is actually a de-optimization for RuntimeMethodInfo with the two changes above, and about neutral for the rest.
- Avoid virtual call for ReflectedType comparison
- Restore comment for RuntimeMethodInfo

jkotas commented:
@buyaa-n I have pushed a commit with a bunch of perf tweaks and simplifications. Let me know if it looks good to you or if you have any questions.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.CoreCLR.cs:
```diff
@@ -118,6 +119,12 @@ public override string ToString()
         }
+
+        public override bool Equals(object? obj) =>
+            obj is RuntimeConstructorInfo ci && m_handle == ci.m_handle && ReferenceEquals(m_declaringType, ci.m_declaringType);
```
```suggestion
            ReferenceEquals(this, obj) ||
            (MetadataUpdater.IsSupported && CacheEquals(obj));
```
Is there a reason why we are not doing this here - so that it looks the same as PropertyInfo and EventInfo?

buyaa-n replied:
Adding `ReferenceEquals(this, obj)` did not improve perf, second have been updating `RuntimeConstructorInfo` same as `RuntimeMethodInfo` as it is a method, but that might wrong

buyaa-n commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.CoreCLR.cs:
```diff
@@ -158,13 +158,17 @@ public override string ToString()
             obj is RuntimeMethodInfo m && m_handle == m.m_handle &&
             ReferenceEquals(m_declaringType, m.m_declaringType) &&
-            ReferenceEquals(ReflectedType, m.ReflectedType);
+            ReferenceEquals(m_reflectedTypeCache.GetRuntimeType(), m_reflectedTypeCache.GetRuntimeType());
```
```suggestion
            ReferenceEquals(m_reflectedTypeCache.GetRuntimeType(), m.m_reflectedTypeCache.GetRuntimeType());
```

buyaa-n committed:
Apply suggestions from code review

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

buyaa-n commented:
> @buyaa-n I have pushed a commit with a bunch of perf tweaks and simplifications. The commit description has the details. Let me know if it looks good to you or if you have any questions.

Thanks for the description, makes sense to me

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeFieldInfo.cs:
```diff
@@ -46,6 +47,13 @@ internal RuntimeType GetDeclaringTypeInternal()
+            (MetadataUpdater.IsSupported && CacheEquals(obj));
+
+        public override int GetHashCode() =>
+            HashCode.Combine(MetadataToken.GetHashCode(), m_declaringType.GetUnderlyingNativeHandle().GetHashCode());
```
The Metadata token property is expensive virtual call here. I have missed this detail. The GetHashCode and Equals should be moved to RtFieldInfo and MdFieldInfo, and changed to directly use the relevant fields - handle field for RtFieldInfo and token field for MdFieldInfo.

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.CoreCLR.cs:
```diff
@@ -118,6 +119,13 @@ public override string ToString()
+
+        public override bool Equals(object? obj) =>
+            ReferenceEquals(this, obj) ||
+            (MetadataUpdater.IsSupported && CacheEquals(obj));
```
I think CacheEquals needs to check the declaring type for this to work. Otherwise, you can get false positives for different generic instantiations of ConstructorInfo

jkotas commented:
Looks good to me, modulo the last two comments

buyaa-n committed:
Apply feedback

buyaa-n commented:
Failures unrelated and known issue, merging
